### ISSUE TITLE
[rllib] rllib not saving Tensorflow AdamOptimizer's momentum and accumulators

### ISSUE THREAD

**Issue by Purpleslz** [ray-project/ray#8412]
### What is the problem?

Ray: 0.8.4
Tensorflow: 1.14

RLLib is using its own checkpoint saving API instead of  tensorflow's, and this API is not saving AdamOptimizer's momentum and accumulators.

When using  AdamOptimizer, the momentum and accumulators (m, v, beta1_power, beta2_power https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/adam.py#L118) is not saved.

### Reproduction (REQUIRED)

Using official example which default optimizer is Adam: https://github.com/ray-project/ray/blob/master/rllib/examples/custom_tf_policy.py

0. add `checkpoint_freq=1` to `tune.run`'s argument.
1. Run example to save checkpoints.
2. add `restore=${checkpoint_path}` to `tune.run` 's argument, restore from one checkpoint and rerun example.

I checked from this line: https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/optimizer.py#L1169

Tensorflow first creates these slot variables, then restore them. If there are no variables to restore, the `deferred_restorations` will be an empty list. These new created slot variables will be zero.

**Comment by ericl** [comment]
What's a good way of saving them? Now that we have deprecated modelv1, perhaps we can standardize on a Keras method cc @sven1977

**Comment by Purpleslz** [comment]
> What's a good way of saving them? Now that we have deprecated modelv1, perhaps we can standardize on a Keras method? cc @sven1977


[Trainer's save method](https://github.com/ray-project/ray/blob/master/rllib/agents/trainer.py#L629) will only save [self._variables in tf policy class](https://github.com/ray-project/ray/blob/master/rllib/policy/tf_policy.py#L229) which will not contain optimizer's accumulators.

I noticed we have implementations about `export_checkpoint`/`export_model` in `policy`, perhaps we should save model checkpoints and training checkpoints separately.

**Comment by ericl** [comment]
Hmm is the right thing to register those variables for saving as well? Actually I thought the optimizer variables used to be saved, maybe this is a regression.

**Comment by Purpleslz** [comment]
Maybe you're right. The "export_checkpoint way" can support more frameworks (e.g. pytorch), torch_policy and tf_policy can use the same interface `export_checkpoint` and the implementation for tf&pytorch is very simple.

I think both ways are sufficient and reasonable to fix this issue, maybe "register optimizers' variable" is a better one. However, I'm not very familiar with tensorflow, I don't know how to do this.

**Comment by sven1977** [comment]
@Purpleslz Could you try this PR here and confirm whether this fixes your issue? I just made the ModelV2 branch behave the same as the ModelV1 one in terms of automatically collecting all vars in the tf-graph.

**Comment by sven1977** [comment]
https://github.com/ray-project/ray/pull/8480

**Comment by Purpleslz** [comment]
This PR #8480 doesn't solve my problem. I'm not sure whether my verification method is correct：
1. I checked the [https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/optimizer.py#L1169](https://github.com/tensorflow/tensorflow/blob/v1.14.0/tensorflow/python/training/optimizer.py#L1169) this line, this line is restoring optimizers' variables. If there are no variables to restore, the deferred_restorations will be an empty list. After the PR, the list is still empty.
2. I think the Adam optimizers' momentum variables will have `Adam` suffix, accumulators will have `Beta` suffix.  I print `self._variables` keys in [https://github.com/ray-project/ray/blob/master/python/ray/experimental/tf_utils.py#L94](https://github.com/ray-project/ray/blob/master/python/ray/experimental/tf_utils.py#L94), there is no such variables.

**Comment by sven1977** [comment]
I don't actually think the problem is ModelV1/V2 related at all.
In tf, optimizer variables are generated after initializing the global vars: `self._sess.run(tf.global_variables_initializer())` and this is happening after we store a Model's vars (V1/V2 doesn't matter here) in `self._variables`, so the optimizer vars were never part of that list. 

When creating a checkpoint, only things in `self._variables` are stored as the Policy's state.
Let's store the optimizer vars separately (I don't think they should be returned in `Policy.get_weights()`) and then do include them in the storing process (`Policy.get_state()`).

**Comment by sven1977** [comment]
Can you try again? This is working now on my end:
```
trainer.train(). # <- e.g. DQN with an Adam optimizer
chkpoint= trainer.save()
...
new_trainer = ...
new_trainer.restore(chkpoint)
p = new_trainer.get_policy()
p._sess.run(p._optimizer._slots)
```

**Comment by sven1977** [comment]
@Purpleslz

**Comment by ericl** [comment]
Yeah, it makes sense to not include them in the policy get_weights, but only in trainer checkpoints.

**Comment by Purpleslz** [comment]
> Can you try again? This is working now on my end:
> 
> ```
> trainer.train(). # <- e.g. DQN with an Adam optimizer
> chkpoint= trainer.save()
> ...
> new_trainer = ...
> new_trainer.restore(chkpoint)
> p = new_trainer.get_policy()
> p._sess.run(p._optimizer._slots)
> ```

I am using ray repo in the master branch, and merged the [#8840](https://github.com/ray-project/ray/pull/8480). 

However, I got this error: 
```
  File "xxxxx/.pyenv/versions/anaconda3-5.0.1/envs/ray/lib/python3.6/site-packages/tensorflow/python/keras/optimizer_v2/optimizer_v2
.py", line 532, in __getattribute__
    return super(OptimizerV2, self).__getattribute__(name)
AttributeError: 'Adam' object has no attribute 'compute_gradients'
```
I think the problem is in this line: https://github.com/ray-project/ray/blob/497493de8018d73baae9795afd126b4256efd134/rllib/policy/tf_policy.py#L721, when I changed this optimzier to `tf.train.AdamOptimizer` the code could run correctly.

It seems that the optimizers' variables could be saved and restored, when will the pull request be merged?

**Comment by sven1977** [comment]
This PR fixes (and tests) saving optimizer state for later restoring.
https://github.com/ray-project/ray/pull/8480

Will be merged later today.

Closing this issue.
